rir/src/compiler/translations/pir_2_rir.cpp

```diff
@@ -26,73 +28,499 @@ namespace pir {
+                        liveRange.begin = pos;
+                        accumulated.erase(accumulated.find(i));
+                    }
+                } while (ip != bb->begin());
+            }
+            assert(pos == 0);
+
+            // Mark everything that is live at the beginning of the BB.
+            auto markLiveEntry = [&](Value* v) {
+                assert(livenessInterval.count(v));
+                auto& liveRange = livenessInterval[v][bb->id];
+                assert(liveRange.live);
+                liveRange.begin = 0;
+            };
+
+            for (auto v : accumulated)
+                markLiveEntry(v);
+            for (auto pi : accumulatedPhiInput)
+                for (auto v : pi.second)
+                    markLiveEntry(v);
+
+            // Merge everything that is live at the beginning of the BB into the
+            // incoming vars of all predecessors
+            //
+            // Phi inputs should only be merged to BB that are successors of the
+            // input BBs
+            auto merge = [&](BB* bb, const std::set<Value*>& live) {
+                auto& liveOut = liveAtEnd[bb];
+                if (!std::includes(liveOut.begin(), liveOut.end(), live.begin(),
+                                   live.end())) {
+                    liveOut.insert(live.begin(), live.end());
+                    todo.insert(bb);
+                }
+            };
+            auto mergePhiInp = [&](BB* bb) {
+                for (auto in : accumulatedPhiInput) {
+                    auto inBB = in.first;
+                    auto inLive = in.second;
+                    if (bb == inBB ||
+                        cfg.transitivePredecessors[bb->id].count(inBB)) {
+                        merge(bb, inLive);
+                    }
+                }
+            };
+            for (auto pre : cfg.predecessors[bb->id]) {
+                if (!liveAtEnd.count(pre)) {
+                    liveAtEnd[pre] = accumulated;
+                    mergePhiInp(pre);
+                    todo.insert(pre);
+                } else {
+                    merge(pre, accumulated);
+                    mergePhiInp(pre);
+                }
+            }
+        }
+
+        if (verbose) {
+            for (auto ll : livenessInterval) {
+                auto& l = ll.second;
+                ll.first->printRef(std::cout);
+                std::cout << " is live : ";
+                for (size_t i = 0; i < bbsSize; ++i) {
+                    if (l[i].live) {
+                        std::cout << i << " [";
+                        std::cout << l[i].begin << ",";
+                        std::cout << l[i].end << "]  ";
+                    }
+                }
+                std::cout << "\n";
+            }
+        }
+    }
+
+    void computeStackAllocation() {
+        Visitor::run(code->entry, [&](BB* bb) {
+            {
+                // Phi at the beginning of BB, all inputs at the end of
```

Comment:
Is this now the case? that all phis are at the beginning? Also, they are still sequential, right? No parallel semantics yet

Reply:
no. but if they happen to be at the beginning, then it works